### PR TITLE
initial smart wallet example setup

### PR DESCRIPTION
**Pull request by hone1er**
Opening as a draft to get feedback

but this should be a working example of coinbase smart wallet integration

https://github.com/wslyvh/nexth/assets/24376928/a8d90964-82f5-47f3-b252-3412867b287c

https://github.com/wslyvh/nexth/assets/24376928/a5481b9e-0e07-4c0e-9767-231b0474d7ae

https://github.com/wslyvh/nexth/assets/24376928/3a270824-2335-4869-b3e7-8b4971b47b9a

Removed the Coinbase connector specific connect and disconnect buttons in favor of telling the user to use the coinbase wallet in web3Modal and added links to smart wallet repo and docs

Screenshot 2024-05-01 at 8 51 06 PM

### CODE REVIEW

**Comment by hone1er** [comment]
Lmao I didn't push the package.json and now I can't get it working again 😭 sad day

File: packages/app/src/app/hooks/useSWWriteContract.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { WriteContractsParameters } from 'viem/experimental'
+import { useEIP5792WalletClient } from './useEIP5792WalletClient'
+import { useState } from 'react'
+
+export function useSWWriteContracts() {
+  const { data: walletClient } = useEIP5792WalletClient()
```
Comment (jxom) [inline packages/app/src/app/hooks/useSWWriteContract.ts:6]:
```suggestion
  const { data: walletClient } = useWalletClient()
```

File: packages/app/src/app/hooks/useSWWriteContract.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  const [id, setId] = useState<string | undefined>(undefined)
+
+  const writeContracts = (parameters: Omit<WriteContractsParameters, 'account' | 'chain'>) => {
+    if (!walletClient) {
+      throw new Error('Wallet client not available')
+    }
+    // TODO better types on useEIP5792WalletClient should give us account and chain "for free"
+    walletClient
+      .writeContracts({
```
Comment (jxom) [inline packages/app/src/app/hooks/useSWWriteContract.ts:15]:
```suggestion
    writeContracts(walletClient, {
```

Can use tree-shakable actions instead. Needs an `import { writeContracts } from 'viem/experimental'`.

File: packages/app/src/app/hooks/useWalletCapabilities.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+function useWalletCapabilities({ chainId }: { chainId?: number }) {
+  const { data: walletClient } = useEIP5792WalletClient()
+  const [capabilities, setCapabilities] = useState<{ [chainId: number]: WalletCapabilities }>({})
+  const [loading, setLoading] = useState(true)
+
+  useEffect(() => {
+    if (walletClient) {
+      walletClient
+        .getCapabilities()
```
Comment (jxom) [inline packages/app/src/app/hooks/useWalletCapabilities.ts:13]: Same suggestion as above.

Comment (jxom) [reply packages/app/src/app/hooks/useWalletCapabilities.ts:13]: You can do this in wagmi from `2.8.0` (it's building now): https://wagmi.sh/react/api/hooks/useCapabilities

Comment (hone1er) [reply packages/app/src/app/hooks/useWalletCapabilities.ts:13]: That was fast! lol I will update to use the wagmi hook 😊

**Comment by wslyvh** [comment]
This is great! Love it

I wonder a little bit if we should add it as an example page or as a more detailed, full example implementation.
My thinking so far is that example/pages should demonstrate basic functionalities (send ether, transfer tokens, etc.) without requiring custom configurations, etc.

More detailed examples like this, which require different packages/configuration/etc. might be easier to maintain as fork or branch. There are a couple of other branches as examples (e.g. ethers instead of viem, using signature-recovery,etc.). We should probably keep a list of active examples on the readme so its easier to find tho.

In this case, I think you could replace the default connect/sign in flow, include a simple contract, NFT minting and provide the full E2E example and functionalities of smart wallets. 

What do you think?

**Comment by hone1er** [comment]
> This is great! Love it
> 
> I wonder a little bit if we should add it as an example page or as a more detailed, full example implementation. My thinking so far is that example/pages should demonstrate basic functionalities (send ether, transfer tokens, etc.) without requiring custom configurations, etc.
> 
> More detailed examples like this, which require different packages/configuration/etc. might be easier to maintain as fork or branch. There are a couple of other branches as examples (e.g. ethers instead of viem, using signature-recovery,etc.). We should probably keep a list of active examples on the readme so its easier to find tho.
> 
> In this case, I think you could replace the default connect/sign in flow, include a simple contract, NFT minting and provide the full E2E example and functionalities of smart wallets.
> 
> What do you think?

Sounds good to me! I'll get working on this soon. I was actually thinking the smart contract example should have a lot more documentation to it so i think breaking it out into it's own branch and getting more detailed is a great idea.

**Comment by wslyvh** [comment]
Hi @hone1er just checking in and see if this is still something you'd like to contribute?
I believe there have been a lot of advancements in smart contract wallets. Would love to see a Nexth example if you're up for it :)

**Comment by hone1er** [comment]
> Hi @hone1er just checking in and see if this is still something you'd like to contribute? I believe there have been a lot of advancements in smart contract wallets. Would love to see a Nexth example if you're up for it :)

Hey @wslyvh ! Yeah my bad i've been working on some buildathon stuff and trying to get freelance work where i can. Lost my job recently lol if no one gets to some smart wallet examples before i get back to this i'd definitely still like to contribute

I've been working with the smart wallet a bit more lately plus wagmi and viem have experimental features already to make working with smart wallets easier so it should be a pretty smooth setup now

**Comment by wslyvh** [comment]
Hi @hone1er I would love to see a proper Smart contract wallet example, but closing the PR for now.. Happy to chat, once you're ready for it :)

**Pull request closed without merging**